lib/spack/spack/util/url.py

```diff
@@ -66,12 +76,54 @@ def parse(url, scheme='file'):
         if isinstance(url, string_types) else url)
 
     (scheme, netloc, path, params, query, _) = url_obj
+
     scheme = (scheme or 'file').lower()
 
+    # This is the first way that a windows path can be parsed.
+    # (The user leaves out the file:// scheme.)
+    #   examples:
+    #     C:\\a\\b\\c
+    #     X:/a/b/c
+    is_windows_path = (len(scheme) == 1 and 'a' <= scheme and scheme <= 'z')
+    if is_windows_path:
+        netloc = scheme.upper() + ':'
+        scheme = 'file'
+
     if scheme == 'file':
-        path = spack.util.path.canonicalize_path(netloc + path)
+        # If the above windows path case did not hold, check the second case.
+        if not is_windows_path:
+            # This is the other way that a windows path can be parsed.
+            # (The user explicitly provides the file:// scheme.)
+            #   examples:
+            #     file://C:\\a\\b\\c
+            #     file://X:/a/b/c
+            is_windows_path = (len(netloc) == 2 and
+                               netloc[1] == ':' and
+                               (('A' <= netloc[0] and netloc[0] <= 'Z') or
+                                ('a' <= netloc[0] and netloc[0] <= 'z')))
```

Comment:
I think a lot of work is being done here because we do the URL parsing on the input string (the `url` parameter) first and then examine the resulting object. Can we instead format the input string appropriately such that we never end up with the confusing scenario that `netloc` actually refers to the drive letter on Windows? I think it would be fine for example if we require that `url` start with one of `[file://, http://, https://, ftp://, s3://]`.